File: crates/breez-sdk/bindings/langs/android/lib/src/main/kotlin/breez_sdk_spark/webln/WebLnController.kt
package breez_sdk_spark.webln

import breez_sdk_spark.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull

/**
 * Callback for enable requests.
 * Called when a website requests WebLN access.
 *
 * @param domain The domain requesting access
 * @return true to allow access, false to deny
 */
typealias OnEnableRequest = suspend (domain: String) -> Boolean

/**
 * Callback for payment requests.
 * Called when a website requests to send a payment.
 *
 * @param invoice The BOLT11 invoice
 * @param amountSats The amount in satoshis
 * @return true to approve payment, false to reject
 */
typealias OnPaymentRequest = suspend (invoice: String, amountSats: Long) -> Boolean

/**
 * Callback for LNURL requests.
 * Called when a website initiates an LNURL flow.
 *
 * @param request The LNURL request details
 * @return User's response with approval and optional amount/comment
 */
typealias OnLnurlRequest = suspend (request: LnurlRequest) -> LnurlUserResponse

/**
 * Represents the type of LNURL request.
 */
enum class LnurlType {
    /** LNURL-pay request */
    PAY,
    /** LNURL-withdraw request */
    WITHDRAW,
    /** LNURL-auth request */
    AUTH
}

/**
 * Represents an LNURL request that needs user approval.
 * Passed to the onLnurlRequest callback.
 */
data class LnurlRequest(
    /** The type of LNURL request */
    val type: LnurlType,
    /** The domain of the LNURL service */
    val domain: String,
    /** Minimum amount in sats (for pay/withdraw requests) */
    val minAmountSats: Long? = null,
    /** Maximum amount in sats (for pay/withdraw requests) */
    val maxAmountSats: Long? = null,
    /** LNURL metadata JSON string (for pay requests) */
    val metadata: String? = null,
    /** Default description (for withdraw requests) */
    val defaultDescription: String? = null
)

/**
 * Represents the user's response to an LNURL request.
 * Returned from the onLnurlRequest callback.
 */
data class LnurlUserResponse(
    /** Whether the user approved the request */
    val approved: Boolean,
    /** Amount in sats selected by the user (for pay/withdraw) */
    val amountSats: Long? = null,
    /** Optional comment (for LNURL-pay) */
    val comment: String? = null
)

/**
 * WebLN error codes returned to JavaScript.
 */
object WebLnErrorCode {
    const val USER_REJECTED = "USER_REJECTED"
    const val PROVIDER_NOT_ENABLED = "PROVIDER_NOT_ENABLED"
    const val INSUFFICIENT_FUNDS = "INSUFFICIENT_FUNDS"
    const val INVALID_PARAMS = "INVALID_PARAMS"
    const val UNSUPPORTED_METHOD = "UNSUPPORTED_METHOD"
    const val INTERNAL_ERROR = "INTERNAL_ERROR"
}

/**
 * Controller for WebLN support in Android WebViews.
 *
 * Injects the WebLN provider JavaScript into WebViews and handles
 * communication between the web page and the Breez SDK.
 *
 * Usage:
 * ```kotlin
 * val webView = WebLnWebView.fromAndroid(androidWebView)
 * val controller = WebLnController(
 *     sdk = sdk,
 *     webView = webView,
 *     onEnableRequest = { domain -> showEnableDialog(domain) },
 *     onPaymentRequest = { invoice, amountSats -> showPaymentDialog(invoice, amountSats) },
 *     onLnurlRequest = { request -> handleLnurlRequest(request) }
 * )
 * controller.inject()
 * ```
 *
 * @param sdk The Breez SDK instance
 * @param webView Platform-specific WebView wrapper
 * @param onEnableRequest Callback when a site requests WebLN access
 * @param onPaymentRequest Callback when a site requests payment approval
 * @param onLnurlRequest Callback when a site initiates an LNURL flow
 */
class WebLnController(
    private val sdk: BreezSdkInterface,
    private val webView: WebLnWebView,
    private val onEnableRequest: OnEnableRequest,
    private val onPaymentRequest: OnPaymentRequest,
    private val onLnurlRequest: OnLnurlRequest,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main)
) {
    private val enabledDomains = mutableSetOf<String>()
    private var cachedPubkey: String? = null
    private val json = Json { ignoreUnknownKeys = true }

    companion object {
        private val SUPPORTED_METHODS = listOf(
            "getInfo", "sendPayment", "makeInvoice",
            "signMessage", "verifyMessage", "lnurl"
        )
    }

    /**
     * Injects the WebLN provider script into the WebView
     */
    fun inject() {
        webView.enableJavaScript()
        webView.addMessageHandler("BreezSparkWebLn") { requestJson ->
            handleMessage(requestJson)
        }

        // Inject the provider script
        webView.injectScript(WEBLN_PROVIDER_SCRIPT)
    }

    /**
     * Cleans up resources. Call when the WebView is destroyed.
     */
    fun dispose() {
        webView.removeMessageHandler("BreezSparkWebLn")
        enabledDomains.clear()
    }

    /**
     * Handles incoming WebLN requests from JavaScript
     */
    private fun handleMessage(requestJson: String) {
        scope.launch {
            try {
                val request = json.parseToJsonElement(requestJson).jsonObject
                val id = request["id"]?.jsonPrimitive?.content ?: return@launch
                val method = request["method"]?.jsonPrimitive?.content ?: return@launch
                val params = request["params"]?.jsonObject ?: JsonObject(emptyMap())

                when (method) {
                    "enable" -> handleEnable(id, params)
                    "getInfo" -> handleGetInfo(id)
                    "sendPayment" -> handleSendPayment(id, params)
                    "makeInvoice" -> handleMakeInvoice(id, params)
                    "signMessage" -> handleSignMessage(id, params)
                    "verifyMessage" -> handleVerifyMessage(id, params)
                    "lnurl" -> handleLnurl(id, params)
                    else -> respond(id, error = "UNSUPPORTED_METHOD")
                }
            } catch (e: Exception) {
                android.util.Log.e("BreezSparkWebLn", "Error handling request", e)
            }
        }
    }

    private suspend fun handleEnable(id: String, params: JsonObject) {
        val domain = params["domain"]?.jsonPrimitive?.content
        if (domain == null) {
            respond(id, error = "INVALID_PARAMS")
            return
        }

        if (enabledDomains.contains(domain)) {
            respond(id, result = JsonObject(emptyMap()))
            return
        }

        val approved = onEnableRequest(domain)
        if (approved) {
            enabledDomains.add(domain)
            respond(id, result = JsonObject(emptyMap()))
        } else {
            respond(id, error = "USER_REJECTED")
        }
    }

    private suspend fun handleGetInfo(id: String) {
        try {
            val pubkey = getNodePubkey()
            respond(id, result = buildJsonObject {
                put("node", buildJsonObject {
                    put("pubkey", JsonPrimitive(pubkey))
                    put("alias", JsonPrimitive(""))
                })
                put("methods", Json.parseToJsonElement(json.encodeToString(SUPPORTED_METHODS)))
            })
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private suspend fun getNodePubkey(): String {
        cachedPubkey?.let { return it }

        val response = sdk.signMessage(
            SignMessageRequest(message = "webln_pubkey_request", compact = true)
        )
        cachedPubkey = response.pubkey
        return response.pubkey
    }

    private suspend fun handleSendPayment(id: String, params: JsonObject) {
        val paymentRequest = params["paymentRequest"]?.jsonPrimitive?.content
        if (paymentRequest == null) {
            respond(id, error = "INVALID_PARAMS")
            return
        }

        try {
            // Parse the invoice to get amount
            val parsed = sdk.parse(paymentRequest)
            var amountSats = 0L

            if (parsed is InputType.Bolt11Invoice) {
                parsed.v1.amountMsat?.let { msat ->
                    amountSats = msat.toLong() / 1000
                }
            }

            // Request payment confirmation from user
            val approved = onPaymentRequest(paymentRequest, amountSats)
            if (!approved) {
                respond(id, error = "USER_REJECTED")
                return
            }

            // Prepare and send payment
            val prepared = sdk.prepareSendPayment(
                PrepareSendPaymentRequest(paymentRequest = paymentRequest)
            )
            val result = sdk.sendPayment(
                SendPaymentRequest(
                    prepareResponse = prepared,
                    options = SendPaymentOptions.Bolt11Invoice(
                        preferSpark = false,
                        completionTimeoutSecs = 60u,
                    ),
                )
            )

            // Extract preimage from payment details
            var preimage = ""
            when (val details = result.payment.details) {
                is PaymentDetails.Lightning -> {
                    preimage = details.htlcDetails.preimage ?: ""
                }
                else -> {}
            }

            respond(id, result = buildJsonObject {
                put("preimage", JsonPrimitive(preimage))
            })
        } catch (e: SdkException.InsufficientFunds) {
            respond(id, error = "INSUFFICIENT_FUNDS")
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private suspend fun handleMakeInvoice(id: String, params: JsonObject) {
        try {
            val amount = params["amount"]?.jsonPrimitive?.longOrNull
                ?: params["defaultAmount"]?.jsonPrimitive?.longOrNull
            val memo = params["defaultMemo"]?.jsonPrimitive?.content ?: ""

            val response = sdk.receivePayment(
                ReceivePaymentRequest(
                    paymentMethod = ReceivePaymentMethod.Bolt11Invoice(
                        description = memo,
                        amountSats = amount?.toULong(),
                        expirySecs = null,
                        paymentHash = null
                    )
                )
            )

            respond(id, result = buildJsonObject {
                put("paymentRequest", JsonPrimitive(response.paymentRequest))
            })
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private suspend fun handleSignMessage(id: String, params: JsonObject) {
        val message = params["message"]?.jsonPrimitive?.content
        if (message == null) {
            respond(id, error = "INVALID_PARAMS")
            return
        }

        try {
            val response = sdk.signMessage(
                SignMessageRequest(message = message, compact = true)
            )
            respond(id, result = buildJsonObject {
                put("message", JsonPrimitive(message))
                put("signature", JsonPrimitive(response.signature))
            })
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private suspend fun handleVerifyMessage(id: String, params: JsonObject) {
        val signature = params["signature"]?.jsonPrimitive?.content
        val message = params["message"]?.jsonPrimitive?.content

        if (signature == null || message == null) {
            respond(id, error = "INVALID_PARAMS")
            return
        }

        try {
            val pubkey = getNodePubkey()
            val response = sdk.checkMessage(
                CheckMessageRequest(
                    message = message,
                    pubkey = pubkey,
                    signature = signature
                )
            )

            if (response.isValid) {
                respond(id, result = JsonObject(emptyMap()))
            } else {
                respond(id, error = "INVALID_PARAMS")
            }
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private suspend fun handleLnurl(id: String, params: JsonObject) {
        val lnurlString = params["lnurl"]?.jsonPrimitive?.content
        if (lnurlString == null) {
            respond(id, error = "INVALID_PARAMS")
            return
        }

        try {
            when (val parsed = sdk.parse(lnurlString)) {
                is InputType.LnurlPay -> handleLnurlPay(id, parsed.v1)
                is InputType.LnurlWithdraw -> handleLnurlWithdraw(id, parsed.v1)
                is InputType.LnurlAuth -> handleLnurlAuth(id, parsed.v1)
                else -> respond(id, error = "INVALID_PARAMS")
            }
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private suspend fun handleLnurlPay(id: String, data: LnurlPayRequestDetails) {
        val lnurlResponse = onLnurlRequest(
            LnurlRequest(
                type = LnurlType.PAY,
                domain = data.domain,
                minAmountSats = data.minSendable.toLong() / 1000,
                maxAmountSats = data.maxSendable.toLong() / 1000,
                metadata = data.metadataStr
            )
        )

        if (!lnurlResponse.approved) {
            respond(id, error = "USER_REJECTED")
            return
        }

        try {
            val prepared = sdk.prepareLnurlPay(
                PrepareLnurlPayRequest(
                    payRequest = data,
                    amountSats = (lnurlResponse.amountSats ?: 0).toULong(),
                    comment = lnurlResponse.comment
                )
            )

            val result = sdk.lnurlPay(
                LnurlPayRequest(prepareResponse = prepared)
            )

            // Extract preimage
            var preimage = ""
            when (val details = result.payment.details) {
                is PaymentDetails.Lightning -> {
                    preimage = details.htlcDetails.preimage ?: ""
                }
                else -> {}
            }

            respond(id, result = buildJsonObject {
                put("status", JsonPrimitive("OK"))
                put("preimage", JsonPrimitive(preimage))
            })
        } catch (e: SdkException.InsufficientFunds) {
            respond(id, error = "INSUFFICIENT_FUNDS")
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private suspend fun handleLnurlWithdraw(id: String, data: LnurlWithdrawRequestDetails) {
        val domain = try {
            java.net.URI(data.callback).host
        } catch (e: Exception) {
            data.callback
        }

        val lnurlResponse = onLnurlRequest(
            LnurlRequest(
                type = LnurlType.WITHDRAW,
                domain = domain,
                minAmountSats = data.minWithdrawable.toLong() / 1000,
                maxAmountSats = data.maxWithdrawable.toLong() / 1000,
                defaultDescription = data.defaultDescription
            )
        )

        if (!lnurlResponse.approved) {
            respond(id, error = "USER_REJECTED")
            return
        }

        try {
            sdk.lnurlWithdraw(
                LnurlWithdrawRequest(
                    withdrawRequest = data,
                    amountSats = (lnurlResponse.amountSats ?: 0).toULong()
                )
            )
            respond(id, result = buildJsonObject {
                put("status", JsonPrimitive("OK"))
            })
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private suspend fun handleLnurlAuth(id: String, data: LnurlAuthRequestDetails) {
        val lnurlResponse = onLnurlRequest(
            LnurlRequest(
                type = LnurlType.AUTH,
                domain = data.domain
            )
        )

        if (!lnurlResponse.approved) {
            respond(id, error = "USER_REJECTED")
            return
        }

        try {
            sdk.lnurlAuth(data)
            respond(id, result = buildJsonObject {
                put("status", JsonPrimitive("OK"))
            })
        } catch (e: Exception) {
            respond(id, error = "INTERNAL_ERROR")
        }
    }

    private fun respond(id: String, result: JsonObject? = null, error: String? = null) {
        val response = buildJsonObject {
            put("id", JsonPrimitive(id))
            put("success", JsonPrimitive(error == null))
            result?.let { put("result", it) }
            error?.let { put("error", JsonPrimitive(it)) }
        }

        val responseJson = json.encodeToString(response)
        webView.post {
            webView.evaluateJavaScript(
                "window.__breezSparkWebLnHandleResponse($responseJson);",
                null
            )
        }
    }
}

File: crates/breez-sdk/bindings/langs/swift/Sources/BreezSdkSpark/WebLn/WebLnController.swift
import Foundation
import WebKit

// MARK: - Callback Type Aliases

/// Callback for enable requests.
/// Called when a website requests WebLN access.
///
/// - Parameter domain: The domain requesting access
/// - Returns: `true` to allow access, `false` to deny
public typealias OnEnableRequest = (String) async -> Bool

/// Callback for payment requests.
/// Called when a website requests to send a payment.
///
/// - Parameters:
///   - invoice: The BOLT11 invoice
///   - amountSats: The amount in satoshis
/// - Returns: `true` to approve payment, `false` to reject
public typealias OnPaymentRequest = (String, Int64) async -> Bool

/// Callback for LNURL requests.
/// Called when a website initiates an LNURL flow.
///
/// - Parameter request: The LNURL request details
/// - Returns: User's response with approval and optional amount/comment
public typealias OnLnurlRequest = (LnurlRequest) async -> LnurlUserResponse

// MARK: - Types

/// Represents the type of LNURL request.
public enum LnurlType {
    /// LNURL-pay request
    case pay
    /// LNURL-withdraw request
    case withdraw
    /// LNURL-auth request
    case auth
}

/// Represents an LNURL request that needs user approval.
/// Passed to the `onLnurlRequest` callback.
public struct LnurlRequest {
    /// The type of LNURL request
    public let type: LnurlType
    /// The domain of the LNURL service
    public let domain: String
    /// Minimum amount in sats (for pay/withdraw requests)
    public let minAmountSats: Int64?
    /// Maximum amount in sats (for pay/withdraw requests)
    public let maxAmountSats: Int64?
    /// LNURL metadata JSON string (for pay requests)
    public let metadata: String?
    /// Default description (for withdraw requests)
    public let defaultDescription: String?

    public init(
        type: LnurlType,
        domain: String,
        minAmountSats: Int64? = nil,
        maxAmountSats: Int64? = nil,
        metadata: String? = nil,
        defaultDescription: String? = nil
    ) {
        self.type = type
        self.domain = domain
        self.minAmountSats = minAmountSats
        self.maxAmountSats = maxAmountSats
        self.metadata = metadata
        self.defaultDescription = defaultDescription
    }
}

/// Represents the user's response to an LNURL request.
/// Returned from the `onLnurlRequest` callback.
public struct LnurlUserResponse {
    /// Whether the user approved the request
    public let approved: Bool
    /// Amount in sats selected by the user (for pay/withdraw)
    public let amountSats: Int64?
    /// Optional comment (for LNURL-pay)
    public let comment: String?

    public init(approved: Bool, amountSats: Int64? = nil, comment: String? = nil) {
        self.approved = approved
        self.amountSats = amountSats
        self.comment = comment
    }
}

/// WebLN error codes returned to JavaScript.
public enum WebLnErrorCode {
    public static let userRejected = "USER_REJECTED"
    public static let providerNotEnabled = "PROVIDER_NOT_ENABLED"
    public static let insufficientFunds = "INSUFFICIENT_FUNDS"
    public static let invalidParams = "INVALID_PARAMS"
    public static let unsupportedMethod = "UNSUPPORTED_METHOD"
    public static let internalError = "INTERNAL_ERROR"
}

// MARK: - WebLnController

/// Controller for WebLN support in iOS WKWebViews.
///
/// Injects the WebLN provider JavaScript into WebViews and handles
/// communication between the web page and the Breez SDK.
///
/// Usage:
/// ```swift
/// let controller = WebLnController(
///     sdk: sdk,
///     webView: webView,
///     onEnableRequest: { domain in
///         return await showEnableDialog(domain)
///     },
///     onPaymentRequest: { invoice, amountSats in
///         return await showPaymentDialog(invoice, amountSats)
///     },
///     onLnurlRequest: { request in
///         return await handleLnurlRequest(request)
///     }
/// )
/// controller.inject()
/// ```
public class WebLnController: NSObject, WKScriptMessageHandler {
    private let sdk: BreezSdk
    private weak var webView: WKWebView?
    private let onEnableRequest: OnEnableRequest
    private let onPaymentRequest: OnPaymentRequest
    private let onLnurlRequest: OnLnurlRequest

    private var enabledDomains = Set<String>()
    private var cachedPubkey: String?

    private static let supportedMethods = [
        "getInfo", "sendPayment", "makeInvoice",
        "signMessage", "verifyMessage", "lnurl"
    ]

    /// Creates a new WebLN controller.
    ///
    /// - Parameters:
    ///   - sdk: The Breez SDK instance
    ///   - webView: The WKWebView to inject WebLN into
    ///   - onEnableRequest: Callback when a site requests WebLN access
    ///   - onPaymentRequest: Callback when a site requests payment approval
    ///   - onLnurlRequest: Callback when a site initiates an LNURL flow
    public init(
        sdk: BreezSdk,
        webView: WKWebView,
        onEnableRequest: @escaping OnEnableRequest,
        onPaymentRequest: @escaping OnPaymentRequest,
        onLnurlRequest: @escaping OnLnurlRequest
    ) {
        self.sdk = sdk
        self.webView = webView
        self.onEnableRequest = onEnableRequest
        self.onPaymentRequest = onPaymentRequest
        self.onLnurlRequest = onLnurlRequest
        super.init()
    }

    /// Injects the WebLN provider script into the WebView
    public func inject() {
        guard let webView = webView else { return }

        // Add script message handler
        webView.configuration.userContentController.add(self, name: "BreezSparkWebLn")

        // Inject provider script
        let script = WKUserScript(
            source: weblnProviderScript,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: false
        )
        webView.configuration.userContentController.addUserScript(script)
    }

    /// WKScriptMessageHandler implementation
    public func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard let body = message.body as? String,
              let data = body.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let id = json["id"] as? String,
              let method = json["method"] as? String else {
            return
        }

        let params = json["params"] as? [String: Any] ?? [:]

        Task {
            await handleMessage(id: id, method: method, params: params)
        }
    }

    private func handleMessage(id: String, method: String, params: [String: Any]) async {
        switch method {
        case "enable":
            await handleEnable(id: id, params: params)
        case "getInfo":
            await handleGetInfo(id: id)
        case "sendPayment":
            await handleSendPayment(id: id, params: params)
        case "makeInvoice":
            await handleMakeInvoice(id: id, params: params)
        case "signMessage":
            await handleSignMessage(id: id, params: params)
        case "verifyMessage":
            await handleVerifyMessage(id: id, params: params)
        case "lnurl":
            await handleLnurl(id: id, params: params)
        default:
            respond(id: id, error: "UNSUPPORTED_METHOD")
        }
    }

    private func handleEnable(id: String, params: [String: Any]) async {
        guard let domain = params["domain"] as? String else {
            respond(id: id, error: "INVALID_PARAMS")
            return
        }

        if enabledDomains.contains(domain) {
            respond(id: id, result: [:])
            return
        }

        let approved = await onEnableRequest(domain)
        if approved {
            enabledDomains.insert(domain)
            respond(id: id, result: [:])
        } else {
            respond(id: id, error: "USER_REJECTED")
        }
    }

    private func handleGetInfo(id: String) async {
        do {
            let pubkey = try await getNodePubkey()
            respond(id: id, result: [
                "node": ["pubkey": pubkey, "alias": ""],
                "methods": Self.supportedMethods
            ])
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func getNodePubkey() async throws -> String {
        if let cached = cachedPubkey {
            return cached
        }

        let response = try await sdk.signMessage(
            request: SignMessageRequest(message: "webln_pubkey_request", compact: true)
        )
        cachedPubkey = response.pubkey
        return response.pubkey
    }

    private func handleSendPayment(id: String, params: [String: Any]) async {
        guard let paymentRequest = params["paymentRequest"] as? String else {
            respond(id: id, error: "INVALID_PARAMS")
            return
        }

        do {
            // Parse the invoice to get amount
            let parsed = try await sdk.parse(input: paymentRequest)
            var amountSats: Int64 = 0

            if case let .bolt11Invoice(details) = parsed {
                if let msat = details.amountMsat {
                    amountSats = Int64(msat / 1000)
                }
            }

            // Request payment confirmation from user
            let approved = await onPaymentRequest(paymentRequest, amountSats)
            guard approved else {
                respond(id: id, error: "USER_REJECTED")
                return
            }

            // Prepare and send payment
            let prepared = try await sdk.prepareSendPayment(
                request: PrepareSendPaymentRequest(paymentRequest: paymentRequest)
            )
            let result = try await sdk.sendPayment(
                request: SendPaymentRequest(
                    prepareResponse: prepared,
                    options: .bolt11Invoice(
                        preferSpark: false,
                        completionTimeoutSecs: 60
                    )
                )
            )

            // Extract preimage from payment details
            var preimage = ""
            if case .lightning(_, _, _, let htlcDetails, _, _, _) = result.payment.details {
                preimage = htlcDetails.preimage ?? ""
            }

            respond(id: id, result: ["preimage": preimage])
        } catch let error as SdkError {
            if case .InsufficientFunds = error {
                respond(id: id, error: "INSUFFICIENT_FUNDS")
            } else {
                respond(id: id, error: "INTERNAL_ERROR")
            }
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func handleMakeInvoice(id: String, params: [String: Any]) async {
        do {
            let amount = (params["amount"] as? Int64) ?? (params["defaultAmount"] as? Int64)
            let memo = params["defaultMemo"] as? String ?? ""

            let response = try await sdk.receivePayment(
                request: ReceivePaymentRequest(
                    paymentMethod: .bolt11Invoice(
                        description: memo,
                        amountSats: amount.map { UInt64($0) },
                        expirySecs: nil,
                        paymentHash: nil
                    )
                )
            )

            respond(id: id, result: ["paymentRequest": response.paymentRequest])
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func handleSignMessage(id: String, params: [String: Any]) async {
        guard let message = params["message"] as? String else {
            respond(id: id, error: "INVALID_PARAMS")
            return
        }

        do {
            let response = try await sdk.signMessage(
                request: SignMessageRequest(message: message, compact: true)
            )
            respond(id: id, result: [
                "message": message,
                "signature": response.signature
            ])
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func handleVerifyMessage(id: String, params: [String: Any]) async {
        guard let signature = params["signature"] as? String,
              let message = params["message"] as? String else {
            respond(id: id, error: "INVALID_PARAMS")
            return
        }

        do {
            let pubkey = try await getNodePubkey()
            let response = try await sdk.checkMessage(
                request: CheckMessageRequest(
                    message: message,
                    pubkey: pubkey,
                    signature: signature
                )
            )

            if response.isValid {
                respond(id: id, result: [:])
            } else {
                respond(id: id, error: "INVALID_PARAMS")
            }
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func handleLnurl(id: String, params: [String: Any]) async {
        guard let lnurlString = params["lnurl"] as? String else {
            respond(id: id, error: "INVALID_PARAMS")
            return
        }

        do {
            let parsed = try await sdk.parse(input: lnurlString)

            switch parsed {
            case let .lnurlPay(details):
                await handleLnurlPay(id: id, data: details)
            case let .lnurlWithdraw(details):
                await handleLnurlWithdraw(id: id, data: details)
            case let .lnurlAuth(details):
                await handleLnurlAuth(id: id, data: details)
            default:
                respond(id: id, error: "INVALID_PARAMS")
            }
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func handleLnurlPay(id: String, data: LnurlPayRequestDetails) async {
        let lnurlResponse = await onLnurlRequest(LnurlRequest(
            type: .pay,
            domain: data.domain,
            minAmountSats: Int64(data.minSendable / 1000),
            maxAmountSats: Int64(data.maxSendable / 1000),
            metadata: data.metadataStr
        ))

        guard lnurlResponse.approved else {
            respond(id: id, error: "USER_REJECTED")
            return
        }

        do {
            let prepared = try await sdk.prepareLnurlPay(
                request: PrepareLnurlPayRequest(
                    amountSats: UInt64(lnurlResponse.amountSats ?? 0),
                    payRequest: data,
                    comment: lnurlResponse.comment
                )
            )

            let result = try await sdk.lnurlPay(
                request: LnurlPayRequest(prepareResponse: prepared)
            )

            // Extract preimage
            var preimage = ""
            if case .lightning(_, _, _, let htlcDetails, _, _, _) = result.payment.details {
                preimage = htlcDetails.preimage ?? ""
            }

            respond(id: id, result: ["status": "OK", "preimage": preimage])
        } catch let error as SdkError {
            if case .InsufficientFunds = error {
                respond(id: id, error: "INSUFFICIENT_FUNDS")
            } else {
                respond(id: id, error: "INTERNAL_ERROR")
            }
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func handleLnurlWithdraw(id: String, data: LnurlWithdrawRequestDetails) async {
        let domain: String
        if let url = URL(string: data.callback) {
            domain = url.host ?? data.callback
        } else {
            domain = data.callback
        }

        let lnurlResponse = await onLnurlRequest(LnurlRequest(
            type: .withdraw,
            domain: domain,
            minAmountSats: Int64(data.minWithdrawable / 1000),
            maxAmountSats: Int64(data.maxWithdrawable / 1000),
            defaultDescription: data.defaultDescription
        ))

        guard lnurlResponse.approved else {
            respond(id: id, error: "USER_REJECTED")
            return
        }

        do {
            _ = try await sdk.lnurlWithdraw(
                request: LnurlWithdrawRequest(
                    amountSats: UInt64(lnurlResponse.amountSats ?? 0),
                    withdrawRequest: data
                )
            )
            respond(id: id, result: ["status": "OK"])
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func handleLnurlAuth(id: String, data: LnurlAuthRequestDetails) async {
        let lnurlResponse = await onLnurlRequest(LnurlRequest(
            type: .auth,
            domain: data.domain
        ))

        guard lnurlResponse.approved else {
            respond(id: id, error: "USER_REJECTED")
            return
        }

        do {
            _ = try await sdk.lnurlAuth(requestData: data)
            respond(id: id, result: ["status": "OK"])
        } catch {
            respond(id: id, error: "INTERNAL_ERROR")
        }
    }

    private func respond(id: String, result: [String: Any]? = nil, error: String? = nil) {
        var response: [String: Any] = [
            "id": id,
            "success": error == nil
        ]
        if let result = result {
            response["result"] = result
        }
        if let error = error {
            response["error"] = error
        }

        guard let data = try? JSONSerialization.data(withJSONObject: response),
              let jsonString = String(data: data, encoding: .utf8) else {
            return
        }

        DispatchQueue.main.async { [weak self] in
            self?.webView?.evaluateJavaScript(
                "window.__breezSparkWebLnHandleResponse(\(jsonString));",
                completionHandler: nil
            )
        }
    }
}
